packages/datadog-plugin-mysql/src/index.js

```diff
@@ -26,85 +33,32 @@ function createWrapQuery (tracer, config) {
         span.setTag('db.name', this.config.database)
       }
 
-      analyticsSampler.sample(span, config.measured)
+      analyticsSampler.sample(span, this.config.measured)
+      this.enter(span, store)
+    })
 
-      const sequence = scope.bind(query, span).apply(this, arguments)
+    this.addSub('apm:mysql:query:end', () => {
+      this.exit()
+    })
 
-      scope.bind(sequence)
+    this.addSub('apm:mysql:query:error', err => {
+      const span = storage.getStore().span
 
-      span.setTag('resource.name', sequence.sql)
-
-      if (sequence._callback) {
-        sequence._callback = wrapCallback(tracer, span, childOf, sequence._callback)
-      } else {
-        sequence.on('end', () => {
-          span.finish()
+      if (err) {
+        span.addTags({
+          'error.type': err.name,
+          'error.msg': err.message,
+          'error.stack': err.stack
         })
       }
+      span.finish()
+    })
 
-      return sequence
-    }
-  }
-}
-
-function createWrapGetConnection (tracer, config) {
-  return function wrapGetConnection (getConnection) {
-    return function getConnectionWithTrace (cb) {
-      const scope = tracer.scope()
-
-      arguments[0] = scope.bind(cb)
-
-      return scope.bind(getConnection).apply(this, arguments)
-    }
+    this.addSub('apm:mysql:query:async-end', () => {
+      const span = storage.getStore().span
+      span.finish()
+    })
```

Comment:
shouldn't we also listen for `:end` in case it throws sync to close out the span?

Reply:
On line 40 I believe there  there is a subscription for 'apm:mysql:query:end' assuming thats what you are referring to

Reply:
@bmeck throws should be caught and sent to `:error`

Reply:
go it, little surprising code flow my bad